src/langchain_google_memorystore_redis/vector_store.py

```diff
@@ -0,0 +1,764 @@
+        self.type = type
+
+
+class VectorIndexConfig(IndexConfig):
+    SUPPORTED_DISTANCE_STRATEGIES = {
+        DistanceStrategy.COSINE,
+        DistanceStrategy.EUCLIDEAN_DISTANCE,
+        DistanceStrategy.MAX_INNER_PRODUCT,
+    }
+
+    def __init__(
+        self,
+        name: str,
+        field_name: str,
+        type: str,
+        distance_strategy: DistanceStrategy,
+        vector_size: int,
+        data_type: str = DEFAULT_DATA_TYPE,
+    ):
+        """
+        Initializes the VectorIndexConfig object.
+
+        Args:
+            name (str): The unique name for the vector index. This name is used to
+                identify and reference the index within the vector storage system.
+            field_name (str): The name of the field in the data structure that contains
+                the vector data to be indexed. This specifies the target data for indexing.
+            type (str): The type of vector index. This parameter determines the indexing
+                algorithm or structure to be used (e.g., "FLAT", "HNSW").
+            distance_strategy (DistanceStrategy): Enum specifying the metric used to
+                calculate the distance or similarity between vectors. Supported strategies
+                include COSINE, EUCLIDEAN_DISTANCE (L2), and MAX_INNER_PRODUCT (IP),
+                influencing how search results are ranked and returned.
+            vector_size (int): The dimensionality of the vectors that will be stored
+                and indexed. All vectors must conform to this specified size.
+            data_type (str, optional): The data type of the vector elements (e.g., "FLOAT32").
+                This specifies the precision and format of the vector data, affecting storage
+                requirements and possibly search performance.
+        """
+        if distance_strategy not in self.SUPPORTED_DISTANCE_STRATEGIES:
+            supported_strategies = ", ".join(
+                [ds.value for ds in self.SUPPORTED_DISTANCE_STRATEGIES]
+            )
+            raise ValueError(
+                f"Unsupported distance strategy: {distance_strategy}. "
+                f"Supported strategies are: {supported_strategies}."
+            )
+
+        super().__init__(name, field_name, type)
+        self.distance_strategy = distance_strategy
+        self.vector_size = vector_size
+        self.data_type = data_type
+
+    @property
+    def distance_metric(self):
+        mapping = {
+            DistanceStrategy.EUCLIDEAN_DISTANCE: "L2",
+            DistanceStrategy.MAX_INNER_PRODUCT: "IP",
+            DistanceStrategy.DOT_PRODUCT: "IP",
+            DistanceStrategy.COSINE: "COSINE",
+        }
+        return mapping[self.distance_strategy]
+
+
+class HNSWConfig(VectorIndexConfig):
+    """
+    Configuration class for HNSW (Hierarchical Navigable Small World) vector indexes.
+    """
+
+    def __init__(
+        self,
+        name: str,
+        field_name=None,
```

Comment:
```suggestion
        field_name: Optional[str] =None,
```

Reply:
Incorporated in the new PR - also I'll use "field_name: str | None = None"